### PR TITLE
major fixes

### PR DESCRIPTION
**Pull request by dividehex**
This is a collection of major changes to the auth0 rules in order to fix various issues.

### CODE REVIEW

**Comment by dividehex** [comment]
This is still a WIP but I would like to start getting feedback on this work.

File: rules/link-users-by-email-with-metadata.js [hunk 1]
```diff
@@ -22,58 +22,87 @@ function linkUsersByEmailWithMetadata(user, context, callback) {
...
+    const mergedProfiles = allResponses.flat();
+
+    return mergedProfiles;
+  }
+
+  const data = searchMultipleEmailCases();
+
+  try {
+    // Ignore non-verified users
+    data = data.filter((u) => u.email_verified);
```
Comment (gcoxmoz) [inline rules/link-users-by-email-with-metadata.js:72]: Can `data` be 0-length?  if so, it falls into the `else` below that expects `data` to be length>=2 and could get exciting.

File: rules/link-users-by-email-with-metadata.js [hunk 2]
```diff
@@ -22,58 +22,87 @@ function linkUsersByEmailWithMetadata(user, context, callback) {
...
-      data = data.filter((u) => u.email_verified);
-
-      if (data.length === 1) {
-        // The user logged in with an identity which is the only one Auth0 knows about
-        // Do not perform any account linking
-        return callback(null, user, context);
-      }
+  }
+
+  // Since email addresses within auth0 are allowed to me mixed case and the /user-by-email search endpoint
```
Comment (gcoxmoz) [inline rules/link-users-by-email-with-metadata.js:32]: "Allowed to _b_e"?

File: rules/force-ldap-logins-over-ldap.js [hunk 3]
```diff
@@ -31,7 +33,8 @@ function forceLDAPLoginsOverLDAP(user, context, callback) {
   // 'ad' is LDAP - Force LDAP users to log with LDAP here
   if (context.connectionStrategy !== 'ad') {
     for (let domain of MOZILLA_STAFF_DOMAINS) {
```
Comment (gcoxmoz) [inline rules/force-ldap-logins-over-ldap.js:35]: deep assumption that `MOZILLA_STAFF_DOMAINS` is an array of all-lowecase, and always will be.
Decently safe assumption, but still is one.  Maybe comment the assumption in the array up top?

File: connections/firefoxaccounts.js [hunk 4]
```diff
@@ -60,6 +60,12 @@ function firefoxAccountsConnection(accessToken, ctx, cb) {
         user_id: id_token.sub,
         picture: p.avatar,
         preferredLanguage: p.locale,
+        // Mozilla accounts (formerly Firefox Accounts), allows mixed case characters in thier email property
```
Comment (floatingatoll) [inline connections/firefoxaccounts.js:63]: *their

File: rules/link-users-by-email-with-metadata.js [hunk 5]
```diff
@@ -22,58 +22,87 @@ function linkUsersByEmailWithMetadata(user, context, callback) {
...
+    return mergedProfiles;
+  }
+
+  const data = searchMultipleEmailCases();
+
+  try {
+    // Ignore non-verified users
+    data = data.filter((u) => u.email_verified);
+
+    if (data.length === 1) {
```
Comment (floatingatoll) [inline rules/link-users-by-email-with-metadata.js:74]: <= -1 would help with Greg's comment above.

File: rules/link-users-by-email-with-metadata.js [hunk 6]
```diff
@@ -99,50 +128,40 @@ function linkUsersByEmailWithMetadata(user, context, callback) {
...
-          context.primaryUser = primaryUser.user_id;
-          context.primaryUserMetadata = primaryUser.user_metadata || {};
-          return callback(null, user, context);
-        });
-      })
-      .catch((err) => {
-        console.log('An unknown error occurred while linking accounts: ' + err);
-        return callback(err);
+    // We no longer keep the user_metadata nor app_metadata from the secondary account
+    // that is being linked.  If the primary account is LDAP, then it's existing
```
Comment (floatingatoll) [inline rules/link-users-by-email-with-metadata.js:132]: *its

**Review by floatingatoll — COMMENTED** [review]
Just a code comments skim for now, but this generally seems reasonable. I'll review logic tomorrow.

File: connections/firefoxaccounts.js [hunk 7]
```diff
@@ -60,6 +60,12 @@ function firefoxAccountsConnection(accessToken, ctx, cb) {
...
+        // I'm adding this comment here in case we want to actually enforce email case conformity in the future
+        // Although, it will need extensive testing to ensure we are breaking anything.  In the meantime, we will
```
Comment (gcoxmoz) [inline connections/firefoxaccounts.js:65]: *aren't breaking.  Ya chaosmonkey.  :)

**Review by gcoxmoz — APPROVED** [review]
I don't think I'm a qualified reviewer on this repo, and you should take my JS reading with a lot of salt, but, I can see where you're going.  Attempting an r+ but I bet it'll be squashed to just an f+

**Review by floatingatoll — APPROVED** [review]
I've read through all of this and it seems familiar and/or fine. (Other reviews also welcome.)

**Review by frosemond — APPROVED** [review]
Nice clean up work!

**Comment by dividehex** [comment]
The current state of this PR has been tested up and down on the dev tenant.

**Review by floatingatoll — APPROVED** [review]

**Review by frosemond — APPROVED** [review]

**Pull request merged**